pkg/gatewayserver/io/ws/ws.go

```diff
@@ -304,7 +305,16 @@ func (s *srv) handleTraffic(w http.ResponseWriter, r *http.Request) (err error)
 					return err
 				}
 			case down := <-conn.Down():
-				concentratorTime, ok := conn.TimeFromTimestampTime(down.GetScheduled().Timestamp)
+				var (
+					concentratorTime scheduling.ConcentratorTime
+					ok               bool
+				)
+				// If absolute time is set, use that for computing xtime. If not, use the concentrator timestamp.
+				if down.GetScheduled().Time != nil {
+					concentratorTime, ok = conn.TimeFromGatewayTime(*ttnpb.StdTime(down.GetScheduled().Time))
```

Comment:
I'm not sure this is correct. Keep in mind that `ScheduleAt` is not just converting the gateway time, but rather allowing some server time leeway too. 

I think the correct way of handling this is to propagate the `ConcentratorTime` from the `ScheduleAt`/`ScheduleAnytime` instead of reconstructing it in the frontends. 

It is lossy otherwise, because the timestamp does not account for rollovers, and you need to basically construct the logic in `ScheduleAt` if you want to account for everything.

Reply:
I think you're right but this is more generic than LBS. We do the same for UDP as well; https://github.com/TheThingsNetwork/lorawan-stack/blob/2d719336fa53d29019e7acee419d5893024a9a09/pkg/ttnpb/udp/translation.go#L396-L401

The bug is here; https://github.com/TheThingsNetwork/lorawan-stack/blob/2d719336fa53d29019e7acee419d5893024a9a09/pkg/gatewayserver/io/io.go#L599-L604. If it's `timestamp` based, we replace the NS timestamp with the emission but for absolute time, we just forward it as is.

Reply:
So while we can indeed propagate the `ConcentratorTime` to the frontend, that works only for LBS. For UDP, we need the abs time there.

My suggestion would be then to offset the `settings.Time` with the calculated emission value. What do you think?

Reply:
In UDP it is actually possible to use the absolute time (`scheduled.Time`). Unless we actually implement class B scheduling, this is not an option for LBS - we need to convert either time, or timestamp, to a concentrator time.

If we convert time, we need to reinvent the logic of `ScheduleAt`.
If we convert timestamp (which we don't even have due to `settings.Timestamp = 0`), we're doing a lossy conversion